fix(pagination): active element incorrect position on first load

arindam1997007 opened the pull request:
Closes #3151 

## 📝 Description

Fixed Active Element positionon on first load, when pagination is rendered initially with `display:hidden`

## ⛳️ Current behavior (updates)

On first load, instead of left arrow button,  **Page 1** active element is displayed

## 🚀 New behavior

Active Element position is fixed even on first load, using IntersectionObserver.

## 💣 Is this a breaking change (Yes/No):

No

## 📝 Additional Information

According to code submitted by users which can be reproduced, issue occurs if pagination component is rendered using `display:hidden` for the first time, and then switching to `display:block`.

Using **IntersectionObserver**, checking for display of pagination componenet in the viewport, and then running the required functions.


https://github.com/nextui-org/nextui/assets/23236141/9c5d1a2e-a06c-49b5-beba-8711b1beca0f



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Bug Fixes**
  - Resolved an issue with the Pagination component where the active element's position was incorrect when `display: hidden` was applied.

- **New Features**
  - Introduced a new `SubTabs` component to manage tabs and content display within the pagination story.
  - Added multiple new SVG icon components: `ChevronDownIcon`, `PlusIcon`, `SearchIcon`, and `VerticalDotsIcon`.
  - Introduced a `TableData` component with advanced table features like filtering, sorting, pagination, and customizable column visibility.

- **Enhancements**
  - Improved the `usePagination` function with better scrolling behavior and visibility checks.
  - Enhanced table data management with predefined columns, status options, and user information for display purposes.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

arindam1997007 commented:
Hi @wingkwong 
Please review when you get the chance.

wingkwong commented on .changeset/tricky-stingrays-relax.md:
```diff
@@ -0,0 +1,5 @@
+"@nextui-org/pagination": patch
+---
+
+Fix Pagination component position of active element, when rendered with display:hidden #3151
```
```
Fix Pagination component position of active element, when rendered with `display:hidden` (#3151)
```

wingkwong commented on packages/components/pagination/package.json:
```diff
@@ -44,6 +44,7 @@
     "@nextui-org/react-utils": "workspace:*",
     "@nextui-org/shared-icons": "workspace:*",
     "@nextui-org/use-pagination": "workspace:*",
+    "@nextui-org/use-intersection-observer": "workspace:*",
```
since you've added this line, please refresh the lock file as well.

wingkwong commented:
will re-review tonite

wingkwong commented on packages/components/pagination/src/use-pagination.ts:
```diff
@@ -221,11 +225,7 @@ export function usePagination(originalProps: UsePaginationProps) {
-
-    const node = map.get(value);
-
+  function scrollTo(node: HTMLElement | undefined, skipAnimation: boolean) {
```
in `scrollTo` function, there is no logic for `node = undefined` case since it always returns in L229. Therefore, rewrite as the following

```diff
- function scrollTo(node: HTMLElement | undefined, skipAnimation: boolean) {
+ function scrollTo(node: HTMLElement, skipAnimation: boolean) {
-   if (!node || !cursorRef.current) return;
+   if (!cursorRef.current) return;
```

then trigger scrollTo when node is not undefined

```diff
- if (activePage && !disableAnimation && isActiveVisible) {
+ if (node && activePage && !disableAnimation && isActiveVisible) {
  scrollTo(node, activePage === activePageRef.current);
}
```

wingkwong reviewed (changes requested):
By using the [sandbox link](https://codesandbox.io/p/devbox/vsw923?migrateFrom=q2h7q8) posted in the original issue with Pagination in your PR, I see the following result - the numbers are okay but the highlight is missing

Your PR:

![image](https://github.com/nextui-org/nextui/assets/35857179/e4be6969-8542-4768-9fef-45e07cf3fed6)

Expected:

![image](https://github.com/nextui-org/nextui/assets/35857179/bea52e69-d7db-4767-b15a-bcfcb0c0d214)

arindam1997007 commented:
> By using the [sandbox link](https://codesandbox.io/p/devbox/vsw923?migrateFrom=q2h7q8) posted in the original issue with Pagination in your PR, I see the following result - the numbers are okay but the highlight is missing
> 
> Your PR:
> 
> ![image](https://private-user-images.githubusercontent.com/35857179/347820692-e4be6969-8542-4768-9fef-45e07cf3fed6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3MDQ2NTksIm5iZiI6MTcyMDcwNDM1OSwicGF0aCI6Ii8zNTg1NzE3OS8zNDc4MjA2OTItZTRiZTY5NjktODU0Mi00NzY4LTlmZWYtNDVlMDdjZjNmZWQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDEzMjU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE0MWE0ODQ3YWE5ZGFiOWUyODRiMzBlZTM4MGIwZjc3YzMyNWMzYjgzNGViOTljNmRmYjQ3YWU3ZmQxMzBiNmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kktibEFammfRaPx73KCV3GSUjVxawhmMImTQWmekHlY)
> 
> Expected:
> 
> ![image](https://private-user-images.githubusercontent.com/35857179/347821458-bea52e69-d7db-4767-b15a-bcfcb0c0d214.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3MDQ2NTksIm5iZiI6MTcyMDcwNDM1OSwicGF0aCI6Ii8zNTg1NzE3OS8zNDc4MjE0NTgtYmVhNTJlNjktZDdkYi00NzY3LWIxNWEtYmNmY2IwYzBkMjE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzExVDEzMjU1OVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkNmY5OTY1Nzk0NDIzYzZhMmQ0NmYwZjhhZDA3NDJmMzExNDhjMmJmNzY1MGM4YmI5MDFjODJkZWRiOTM4MmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Btn1dtlfNWkCqIUHho4D-cLyM2o3Az-5ZNbPutdDn9w)

Hi @wingkwong where did you test it? In your local machine? Can you tell me which browser and version of browser you are using?

wingkwong commented:
@arindam1997007 I cloned the sandbox code to my local and then replaced pagination import to local one.

arindam1997007 commented:
> @arindam1997007 I cloned the sandbox code to my local and then replaced pagination import to local one. Tested in Chrome.

Okay. I tested it in local Chrome as well. It is working. 
I copied the code from the codesandbox, and put it in `pagination.stories.tsx` and then rendered it.

@wingkwong Is it okay if I publish the `pagination component` which is forked by me in `npm`? And then test it in codesandbox to debug it. I will then delete the package after testing it. Or it will violate your policies?

Otherwise is there any other ways to debug it?

wingkwong commented:
Can you share your modified `pagination.stories.tsx` to me?

arindam1997007 commented:
> Can you share your modified `pagination.stories.tsx` to me? You may commit it and rollback later on.

Sure. I am committin it.

arindam1997007 commented:
> Can you share your modified `pagination.stories.tsx` to me? You may commit it and rollback later on.

@wingkwong I have committed the `pagination.stories` changes.
Need to `pnpm i` before running. I have removed `pagination` dependency from `table` component for this testing purpose, since it was giving a cyclic dependency error.

wingkwong commented:
1. The difference is if i cloned and linked locally, `.data-[moving]opacity-100[data-moving]` is missing
2. if you change `initialPage` to `2`, after switching tab, the default page is still `1`.

arindam1997007 commented:
> 1. The difference is if i cloned and linked locally, `.data-[moving]opacity-100[data-moving]` is missing. And the logic happens to be `scrollTo` you've changed.
> 2. if you change `initialPage` to `2`, after switching tab, the default page is still `1`.

Okay, I am checking this.

wingkwong commented on packages/components/pagination/package.json:
```diff
@@ -44,6 +44,15 @@
     "@nextui-org/shared-icons": "workspace:*",
     "@nextui-org/use-pagination": "workspace:*",
+    "@nextui-org/use-intersection-observer": "workspace:*",
+    "@nextui-org/tabs": "workspace:*",
+    "@nextui-org/table": "workspace:*",
+    "@nextui-org/input": "workspace:*",
+    "@nextui-org/button": "workspace:*",
+    "@nextui-org/dropdown": "workspace:*",
+    "@nextui-org/chip": "workspace:*",
+    "@nextui-org/user": "workspace:*",
```
double check these dependencies. if they are for the stories / tests, put them to `devDependencies` instead.

wingkwong commented on packages/components/pagination/src/use-pagination.ts:
```diff
@@ -112,6 +113,12 @@ interface Props extends Omit<HTMLNextUIProps<"nav">, "onChange"> {
+  /**
+   * Ref to the <ul> container element.
+   */
+  ulElemRef?: RefObject<HTMLElement>;
```
`HTMLUListElement`?

wingkwong commented on packages/components/pagination/src/use-pagination.ts:
```diff
@@ -281,12 +292,21 @@ export function usePagination(originalProps: UsePaginationProps) {
   useEffect(() => {
-    if (activePage && !disableAnimation) {
+    // Setting intersectionObserver to ulELement
+    if (!!ulElemRef?.current) activeIntersectionRef(ulElemRef.current);
```
`!!` is not necessary.

wingkwong commented on packages/components/table/package.json:
```diff
@@ -59,8 +59,7 @@
     "@nextui-org/system": "workspace:*",
     "@nextui-org/button": "workspace:*",
-    "@nextui-org/chip": "workspace:*",
-    "@nextui-org/pagination": "workspace:*",
```
why is this line removed? this package is used in `table.stories.tsx`.

wingkwong commented on packages/components/pagination/stories/table/utils.ts:
```diff
@@ -0,0 +1,3 @@
+export function capitalize(str: string) {
+  return str.charAt(0).toUpperCase() + str.slice(1);
+}
```
use the one from `packages/utilities/shared-utils/src/functions.ts`

wingkwong commented on packages/components/pagination/stories/pagination.stories.tsx:
```diff
@@ -70,10 +73,52 @@ const defaultProps = {
   args: {
     ...defaultProps,
   },
+  render: SubTabs,
```
create a new story instead of modifying the default one.

wingkwong reviewed (commented):
storybook deployment failed. will leave comments for storybook part once it is ready.

wingkwong commented:
The failed deployment was due to you removed the pagination dependencies from table package.

arindam1997007 commented:
> The failed deployment was due to you removed the pagination dependencies from table package.

Yeah I removed it because Turborepo was throwing a cyclic dependency error.
Because, `pagination` is using the `table` component as a dependency.

And `table` component was using `pagination` as a dependency. 

So, only for this testing this particular use case, I removed the pagination dependency from Table.

arindam1997007 commented:
> second thought - do you really need the table? does the issue only reproduce when using table + pagination?

Hey @wingkwong 

The issue can be reproduced without using table/tabs as well. The issue arises when `<Pagination>` component is initially rendered with `display: hidden` and later made visible in the UI.

I have two doubts
1. Should I create a new story specifically for this issue that will be deployed? Or will the story only be used for testing purposes at the moment?
2. If the story is created for deployment, what should be the title of the story?

wingkwong commented:
For testing only so you can just make it as minimal as possible. Ideally there should be a test covering the logic of the fix.

arindam1997007 commented:
> For testing only so you can just make it as minimal as possible. Ideally there should be a test covering the logic of the fix.

Hi @wingkwong 

I have made changes to create a simple story to illustrate the issue.
I tried creating a test covering the logic, but I am unable to get the selected element's (which is a span) inline `transform` style when running the tests. `element.style.transform` is coming as empty and I am not sure why.

Test code:

```js
it("the pagination cursor should be positioned properly", async () => {
    const wrapper = render(
      <Pagination
        {...pagination.defaultVariants}        
        showControls
        initialPage={4}
        total={10}
      />,
    );

    const paginationElement = wrapper.container.firstChild as HTMLElement;

    if (paginationElement) {

      const highlightedSpan = paginationElement.querySelector(
         'span[aria-hidden="true"]',
       ) as HTMLElement;

      console.log(
        highlightedSpan.outerHTML,
        highlightedSpan.innerHTML,
        highlightedSpan.style.transform,
      );     
    }
  });
});
```

Test Output ( `transform` is not present in the outerHTML ) :
```
highlightedSpan.outerHTML:

'<span aria-hidden="true" data-slot="cursor" data-testid="span-highlight" class="absolute flex overflow-visible items-center justify-center origin-center left-0 select-none touch-none pointer-events-none z-20 data-[moving=true]:transition-transform !data-[moving=true]:duration-300 opacity-0 data-[moving]:opacity-100 shadow-primary/40 bg-primary text-primary-foreground shadow-md min-w-9 w-9 h-9 text-small rounded-medium">4</span>'

highlightedSpan.innerHTML:

'4'

highlightedSpan.style.transform:

''
```

wingkwong commented on packages/components/table/package.json:
```diff
@@ -59,8 +59,8 @@
     "@nextui-org/button": "workspace:*",
-    "@nextui-org/chip": "workspace:*",
-    "@nextui-org/pagination": "workspace:*",
+    "@nextui-org/chip": "workspace:*",   
+    "@nextui-org/pagination": "workspace:*", 
```
should not have such change (i.e. adding spaces)

wingkwong reviewed (changes requested):
tests are missing

wingkwong commented:
> > For testing only so you can just make it as minimal as possible. Ideally there should be a test covering the logic of the fix.
> 
> Hi @wingkwong
> 
> I have made changes to create a simple story to illustrate the issue. I tried creating a test covering the logic, but I am unable to get the selected element's (which is a span) inline `transform` style when running the tests. `element.style.transform` is coming as empty and I am not sure why.
> 
> Test code:
> 
> ```js
> it("the pagination cursor should be positioned properly", async () => {
>     const wrapper = render(
>       <Pagination
>         {...pagination.defaultVariants}        
>         showControls
>         initialPage={4}
>         total={10}
>       />,
>     );
> 
>     const paginationElement = wrapper.container.firstChild as HTMLElement;
> 
>     if (paginationElement) {
> 
>       const highlightedSpan = paginationElement.querySelector(
>          'span[aria-hidden="true"]',
>        ) as HTMLElement;
> 
>       console.log(
>         highlightedSpan.outerHTML,
>         highlightedSpan.innerHTML,
>         highlightedSpan.style.transform,
>       );     
>     }
>   });
> });
> ```
> 
> Test Output ( `transform` is not present in the outerHTML ) :
> 
> ```
> highlightedSpan.outerHTML:
> 
> '<span aria-hidden="true" data-slot="cursor" data-testid="span-highlight" class="absolute flex overflow-visible items-center justify-center origin-center left-0 select-none touch-none pointer-events-none z-20 data-[moving=true]:transition-transform !data-[moving=true]:duration-300 opacity-0 data-[moving]:opacity-100 shadow-primary/40 bg-primary text-primary-foreground shadow-md min-w-9 w-9 h-9 text-small rounded-medium">4</span>'
> 
> highlightedSpan.innerHTML:
> 
> '4'
> 
> highlightedSpan.style.transform:
> 
> ''
> ```

@arindam1997007  The logic in your test is incomplete. What are you trying to do with `highlightedSpan.style.transform`? You may explore `getComputedStyle` if you want to get the value of the style.

wingkwong commented:
@arindam1997007 any update?

arindam1997007 commented:
Hi @wingkwong 
I tried but wasn't able to complete the tests for this. 
If the tests are important, you may reassign the task. 
Sorry for the holdup.

wingkwong commented:
resolved in #4222
